<a href="https://colab.research.google.com/github/sophia-moore/232-Final-Project/blob/main/code_regressions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predictive Regressions

In [ ]:
# Import required libraries
import pandas as pd
import numpy as np
import pickle
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from typing import Sequence, Dict, Tuple, Optional
from collections import Counter
from typing import Tuple, Union

In [ ]:
def read_csv_to_df(file_path, index_col=None):
	"""
	Read a CSV file and convert it to a pandas DataFrame

	Args:
		file_path (str): Path to the CSV file

	Returns:
		pandas.DataFrame: The loaded DataFrame
	"""
	df = pd.read_csv(file_path, index_col=index_col)
	return df


In [ ]:
# SIMPLE AVG
Z_train_simple_pca = read_csv_to_df('Z_train_pca_glove_avg.csv', index_col=0)
Z_test_simple_pca = read_csv_to_df('Z_test_pca_glove_avg.csv', index_col=0)

Z_train_simple_lap = read_csv_to_df('Z_train_lap_avg.csv')
Z_test_simple_lap = read_csv_to_df('Z_test_lap_avg.csv')

Z_train_simple_diff = read_csv_to_df('Z_train_diff_t2_avg.csv', index_col=0)
Z_test_simple_diff = read_csv_to_df('Z_test_diff_t2_avg.csv', index_col=0)

print("printing SIMPLE AVG:", Z_train_simple_pca.shape, Z_test_simple_pca.shape, Z_train_simple_lap.shape, Z_test_simple_lap.shape, Z_train_simple_diff.shape, Z_test_simple_diff.shape)

# TFIDF
Z_train_tfidf_pca = read_csv_to_df('Z_train_pca_glove_tfidf.csv', index_col=0)
Z_test_tfidf_pca = read_csv_to_df('Z_test_pca_glove_tfidf.csv', index_col=0)

Z_train_tfidf_lap = read_csv_to_df('Z_train_lap_tfidf.csv')
Z_test_tfidf_lap = read_csv_to_df('Z_test_lap_tfidf.csv')

Z_train_tfidf_diff = read_csv_to_df('Z_train_diff_t2_tfidf.csv', index_col=0)
Z_test_tfidf_diff = read_csv_to_df('Z_test_diff_t2_tfidf.csv', index_col=0)

print("printing TFIDF:", Z_train_tfidf_pca.shape, Z_test_tfidf_pca.shape, Z_train_tfidf_lap.shape, Z_test_tfidf_lap.shape, Z_train_tfidf_diff.shape, Z_test_tfidf_diff.shape)

printing SIMPLE AVG: (8378, 10) (2793, 10) (8378, 10) (2793, 10) (8378, 10) (2793, 10)
printing TFIDF: (8378, 10) (2793, 10) (8378, 10) (2793, 10) (8378, 10) (2793, 10)


In [ ]:
Z_train_tfidf_lap.head(10)

0         1         2         3         4         5         6  \
0  0.008801 -0.007235  0.005716 -0.022118 -0.000451  0.019778  0.011677   
1  0.011468  0.006459 -0.000006 -0.004770 -0.002964 -0.024045 -0.000303   
2 -0.000141 -0.012624  0.017270 -0.001190 -0.000807  0.013633  0.013879   
3 -0.002734  0.007075 -0.001494 -0.021819 -0.005781  0.008696 -0.000912   
4 -0.016654  0.011078  0.010660  0.006292 -0.009847  0.001006 -0.001830   
5 -0.004348 -0.013229  0.011798  0.009824 -0.003068 -0.006986  0.007642   
6  0.005103 -0.005451 -0.001848  0.018525  0.008415  0.008715 -0.021363   
7  0.019075  0.010531 -0.010644  0.014686 -0.011772  0.006127 -0.002948   
8 -0.001012  0.002481  0.003504 -0.006976  0.008647  0.002681  0.000433   
9 -0.005888  0.008640 -0.022148  0.023886  0.025965  0.008209  0.024293   

          7         8         9  
0  0.006789 -0.007947  0.015674  
1  0.004872 -0.001789 -0.005717  
2 -0.004942  0.010545  0.009374  
3  0.000977  0.008981 -0.009696  
4  0.005391  0.004360 -0.010023  
5  0.000893 -0.007043 -0.006425  
6  0.001045  0.005980 -0.005855  
7  0.009185 -0.000132 -0.012027  
8  0.008719  0.003130  0.005382  
9 -0.007008  0.012394 -0.001997

In [ ]:
# Read the pickle file
file_path = r"C:\Users\gpapa\OneDrive\My Life\Education\2025 YALE MMS\Term_4S\T4_E_Adv_Lin_Algebra_p2\final_project\DJN_2017-01.pkl"
with open(file_path, 'rb') as f:
    data = pickle.load(f)

# Display basic information about the dataset
print(type(data))
if isinstance(data, pd.DataFrame):
    print(data.head())
    print(data.info())

<class 'pandas.core.frame.DataFrame'>
       Date   Company                                              Title  \
0  20170101  [ATC.AE]  [press, release, murphy, owner, kxly, abc, cha...   
1  20170101    [ARKR]  [ark, restaurant, apos, ceo, weinstein, result...   
2  20170101       [F]  [like, futurist, be, prepare, totally, unexpec...   
3  20170101    [NCOM]  [press, release, national, commerce, corporati...   
4  20170101     [EDE]  [press, release, district, electric, company, ...   

                                             Article  Compound_Return  \
0  [remove, it, programming, wire, despite, willi...              NaN   
1                             [from, seek, earnings]         0.070817   
2  [by, in, resident, futurist, lead, team, imagi...         0.076860   
3  [part, family, globe, ncc, parent, company, he...        -0.020161   
4  [wire, closing, today, merger, company, subsid...              NaN   

   Day1_Return  
0          NaN  
1     0.008468  
2     0.046068 

C:\Users\gpapa\AppData\Local\Temp\ipykernel_27504\1126090839.py:4: DeprecationWarning: numpy.core.numeric is deprecated and has been renamed to numpy._core.numeric. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.numeric._frombuffer.
  data = pickle.load(f)


In [ ]:

# df = pd.read_csv("path/to/your/dataset.csv")   # <- uncomment & adapt if you haven’t loaded it yet
assert {'Article', 'Day1_Return'}.issubset(data.columns), "Missing one of the required columns."

# 3. Keep only observations with a target value
clean_df = data.dropna(subset=['Day1_Return']).reset_index(drop=True)

base_unique = clean_df.drop_duplicates(subset="Article").copy()
# create a hashable key
base_unique["Article_key"] = base_unique["Article"].apply(tuple)

In [ ]:
# ──────────────────────────────────────────────────────────────────────────
# 1. SPLIT: preserve indices
# ──────────────────────────────────────────────────────────────────────────
X = base_unique["Article_key"]       # use the tuple key directly
y = base_unique["Day1_Return"]
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42, shuffle=True
)
# note: X_train.index and X_test.index are subsets of base_unique.index

In [ ]:
X_train.head()

2083     (operate, direct, insider, surrender, share, b...
10443    (more, follow, press, release, market, drive, ...
18941    (investigate, nyse, so, globe, today, announce...
3578     (salesforcecom, inc, form, file, reflect, inte...
3203     (wire, today, announce, former, member, member...
Name: Article_key, dtype: object

In [ ]:
# ──────────────────────────────────────────────────────────────────────────
# 2. EMBED & DECOMPOSE (PCA example shown)
# ──────────────────────────────────────────────────────────────────────────
# (you already have Z_train_pca and Z_test_pca as NumPy arrays of shape
#  (len(X_train), 10) and (len(X_test), 10), respectively)

def make_aligned_dfs(Ztr: np.ndarray,
                     Zte: np.ndarray,
                     train_idx: pd.Index,
                     test_idx: pd.Index,
                     prefix: str,
                     k: int = 10
) -> tuple[pd.DataFrame, pd.DataFrame]:
    """
    Turn train/test factor arrays into indexed DataFrames with named cols.

    Ztr: (n_train × k) factor array
    Zte: (n_test  × k) factor array
    train_idx: Index for train rows
    test_idx : Index for test  rows
    prefix   : e.g. "pca", "lap", "diff"
    k        : number of components
    """
    cols = [f"{prefix}_{i+1}" for i in range(k)]
    df_tr = pd.DataFrame(Ztr, index=train_idx, columns=cols)
    df_te = pd.DataFrame(Zte, index=test_idx,  columns=cols)
    return df_tr, df_te

# ————————————————————————————————————————————————————————————————
# simple‑avg sets
# ————————————————————————————————————————————————————————————————
Ztr_avg_pca,  Zte_avg_pca  = make_aligned_dfs(Z_train_simple_pca.values,  Z_test_simple_pca.values,
                                          X_train.index, X_test.index, "pca")
Ztr_avg_lap,  Zte_avg_lap  = make_aligned_dfs(Z_train_simple_lap.values,  Z_test_simple_lap.values,
                                          X_train.index, X_test.index, "lap")
Ztr_avg_diff, Zte_avg_diff = make_aligned_dfs(Z_train_simple_diff.values, Z_test_simple_diff.values,
                                          X_train.index, X_test.index, "diff")

# ————————————————————————————————————————————————————————————————
# tfidf sets
# ————————————————————————————————————————————————————————————————
Ztr_tfidf_pca,  Zte_tfdidf_pca  = make_aligned_dfs(Z_train_tfidf_pca.values,  Z_test_tfidf_pca.values,
                                          X_train.index, X_test.index, "pca")
Ztr_tfidf_lap,  Zte_tfidf_lap  = make_aligned_dfs(Z_train_tfidf_lap.values,  Z_test_tfidf_lap.values,
                                          X_train.index, X_test.index, "lap")
Ztr_tfidf_diff, Zte_tidf_diff = make_aligned_dfs(Z_train_tfidf_diff.values, Z_test_tfidf_diff.values,
                                          X_train.index, X_test.index, "diff")

In [ ]:
# only keep **X** PCs for each method

ArrayLike = Union[pd.DataFrame, np.ndarray]

def slice_components(
    Ztr: ArrayLike,
    Zte: ArrayLike,
    n_components: int
) -> Tuple[ArrayLike, ArrayLike]:
    """
    Keep only the first `n_components` columns of Ztr and Zte.
    Works for either DataFrames or ndarrays.

    Returns
    -------
    (Ztr_sliced, Zte_sliced)
    """
    if isinstance(Ztr, pd.DataFrame) and isinstance(Zte, pd.DataFrame):
        return Ztr.iloc[:, :n_components], Zte.iloc[:, :n_components]
    else:
        # assume NumPy arrays
        return Ztr[:, :n_components], Zte[:, :n_components]

# USAGE -- KEEP 3 PCs
Ztr_avg_pca,  Zte_avg_pca  = slice_components(Ztr_avg_pca,  Zte_avg_pca,  3)
Ztr_avg_lap,  Zte_avg_lap  = slice_components(Ztr_avg_lap,  Zte_avg_lap,  3)
Ztr_avg_diff, Zte_avg_diff = slice_components(Ztr_avg_diff, Zte_avg_diff, 3)

Ztr_tfidf_pca,  Zte_tfdidf_pca  = slice_components(Ztr_tfidf_pca,  Zte_tfdidf_pca,  3)
Ztr_tfidf_lap,  Zte_tfidf_lap  = slice_components(Ztr_tfidf_lap,  Zte_tfidf_lap,  3)
Ztr_tfidf_diff, Zte_tidf_diff = slice_components(Ztr_tfidf_diff, Zte_tidf_diff, 3)

In [ ]:
# Reconstruct the full base_unique embedding matrix by concatenating train + test
Z_avg_base_pca  = pd.concat([Ztr_avg_pca,  Zte_avg_pca], axis=0)
Z_avg_base_lap  = pd.concat([Ztr_avg_lap,  Zte_avg_lap], axis=0)
Z_avg_base_diff = pd.concat([Ztr_avg_diff, Zte_avg_diff], axis=0)
# Now Z_simple_base_pca.index == base_unique.index
# and Z_simple_base_pca[pca_cols] holds the 10-d vectors for each unique-article

Z_tfidf_base_pca  = pd.concat([Ztr_tfidf_pca,  Zte_tfdidf_pca], axis=0)
Z_tfidf_base_lap  = pd.concat([Ztr_tfidf_lap,  Zte_tfidf_lap], axis=0)
Z_tfidf_base_diff = pd.concat([Ztr_tfidf_diff, Zte_tidf_diff], axis=0)




In [ ]:
Z_avg_base_pca.head()

pca_1     pca_2     pca_3
2083  -0.643635 -0.264756  0.203019
10443 -0.344426  0.243452  0.094582
18941 -0.052609 -0.086338  0.670755
3578  -0.202244 -0.010270  0.184002
3203   0.605485  0.160494  0.247221

In [ ]:
# ──────────────────────────────────────────────────────────────────────────
# 4. MERGE back to clean_df
# ──────────────────────────────────────────────────────────────────────────

def broadcast_embeddings(
    clean_df: pd.DataFrame,
    base_unique: pd.DataFrame,
    Z_base: pd.DataFrame,
    prefix: str
) -> pd.DataFrame:
    """
    Attach the embedding columns in Z_base (indexed by base_unique.index)
    back onto every row of clean_df that shares the same Article_key.

    clean_df     : full DataFrame with duplicates
    base_unique  : deduped DataFrame with Article_key column
    Z_base       : DataFrame of embeddings, indexed the same as base_unique
    prefix       : e.g. 'avg_pca', 'avg_lap', 'avg_diff',
                   'tfidf_pca', etc.  Used only in warnings.
    """
    # 1️⃣  ensure the key exists
    if "Article_key" not in clean_df.columns:
        clean_df = clean_df.copy()
        clean_df["Article_key"] = clean_df["Article"].apply(tuple)

    # 2️⃣  build a lookup: Article_key → embeddings
    lookup = base_unique[["Article_key"]].join(Z_base)

    # 3️⃣  merge onto clean_df
    merged = clean_df.merge(
        lookup,
        on="Article_key",
        how="left",
        validate="many_to_one"  # each clean_df row matches exactly one lookup row
    )

    # 4️⃣  drop the helper key
    return merged.drop(columns="Article_key")


# —————————————————————————————————————————————————————————————————————
# Now call it for your six embedding bases
# —————————————————————————————————————————————————————————————————————

# Simple‑avg
clean_with_Z_avg_pca  = broadcast_embeddings(clean_df, base_unique, Z_avg_base_pca,  "avg_pca")
clean_with_Z_avg_lap  = broadcast_embeddings(clean_df, base_unique, Z_avg_base_lap,  "avg_lap")
clean_with_Z_avg_diff = broadcast_embeddings(clean_df, base_unique, Z_avg_base_diff, "avg_diff")

# TF‑IDF
clean_with_Z_tf_pca  = broadcast_embeddings(clean_df, base_unique, Z_tfidf_base_pca,  "tfidf_pca")
clean_with_Z_tf_lap  = broadcast_embeddings(clean_df, base_unique, Z_tfidf_base_lap,  "tfidf_lap")
clean_with_Z_tf_diff = broadcast_embeddings(clean_df, base_unique, Z_tfidf_base_diff, "tfidf_diff")


In [ ]:
clean_with_Z_tf_lap.head(5)

Date Company                                              Title  \
0  20170101  [ARKR]  [ark, restaurant, apos, ceo, weinstein, result...   
1  20170101     [F]  [like, futurist, be, prepare, totally, unexpec...   
2  20170101  [NCOM]  [press, release, national, commerce, corporati...   
3  20170101  [SCOR]  [press, release, comscore, report, north, movi...   
4  20170102  [TWTR]  [late, twitter, exec, depart, great, china, he...   

                                             Article  Compound_Return  \
0                             [from, seek, earnings]         0.070817   
1  [by, in, resident, futurist, lead, team, imagi...         0.076860   
2  [part, family, globe, ncc, parent, company, he...        -0.020161   
3  [comscore, report, north, hit, two, last, prne...         0.023541   
4  [say, depart, position, manage, director, late...         0.028676   

   Day1_Return     lap_1     lap_2     lap_3  
0     0.008468  0.000073 -0.000025  0.000009  
1     0.046068 -0.005563 -0.015160  0.001455  
2    -0.004098  0.001350 -0.012502  0.017412  
3     0.025794 -0.000063  0.000015 -0.000020  
4     0.025547 -0.001796 -0.016530 -0.000958

In [ ]:
# Sanity check: make sure the indexes are aligned

def check_index_alignment(base_unique: pd.DataFrame,
                          embedding_bases: dict[str, pd.DataFrame]
) -> None:
    """
    For each named embedding base, print any index values that appear
    in base_unique but not in the embedding, and vice versa.

    embedding_bases: mapping from a string name (e.g. "avg_pca") to the
                     corresponding Z_base DataFrame (indexed by base_unique).
    """
    for name, Z_base in embedding_bases.items():
        missing_in_Z = base_unique.index.difference(Z_base.index)
        extra_in_Z   = Z_base.index.difference(base_unique.index)
        print(f"\n=== Alignment check for {name} ===")
        if missing_in_Z.empty and extra_in_Z.empty:
            print("✔️  Perfect alignment: no index mismatches.")
        else:
            if not missing_in_Z.empty:
                print(f"❌ Indices in base_unique but not in {name}: {missing_in_Z.tolist()}")
            if not extra_in_Z.empty:
                print(f"❌ Indices in {name} but not in base_unique: {extra_in_Z.tolist()}")

# —————————————————————————————————————————————————————————————————————
# Call it once with all three bases
# —————————————————————————————————————————————————————————————————————
check_index_alignment(
    base_unique,
    {
        "avg_pca":  Z_avg_base_pca,
        "avg_lap":  Z_avg_base_lap,
        "avg_diff": Z_avg_base_diff,
        "tfidf_pca":  Z_tfidf_base_pca,
        "tfidf_lap":  Z_tfidf_base_lap,
        "tfidf_diff": Z_tfidf_base_diff,
    }
)



=== Alignment check for avg_pca ===
✔️  Perfect alignment: no index mismatches.

=== Alignment check for avg_lap ===
✔️  Perfect alignment: no index mismatches.

=== Alignment check for avg_diff ===
✔️  Perfect alignment: no index mismatches.

=== Alignment check for tfidf_pca ===
✔️  Perfect alignment: no index mismatches.

=== Alignment check for tfidf_lap ===
✔️  Perfect alignment: no index mismatches.

=== Alignment check for tfidf_diff ===
✔️  Perfect alignment: no index mismatches.


### STEP 2 - RUN REGRESSIONS

In [ ]:
from sklearn.linear_model import Ridge, LinearRegression
from sklearn.metrics       import r2_score, mean_squared_error


def get_train_test_Z(clean_with_Z: pd.DataFrame,
                     prefix: str,
                     X_train_idx: pd.Index,
                     X_test_idx: pd.Index
) -> tuple[np.ndarray, np.ndarray]:
    """
    Extract train/test factor matrices from a broadcasted DataFrame.

    clean_with_Z   : DataFrame with columns prefix_1…prefix_k
    prefix         : e.g. 'avg_pca', 'avg_lap', etc.
    X_train_idx    : Index for training rows
    X_test_idx     : Index for test rows

    Returns
    -------
    Ztr, Zte : NumPy arrays of shape (len(X_train), k), (len(X_test), k)
    """
    # find all columns that start with the prefix
    cols = [c for c in clean_with_Z.columns if c.startswith(prefix + "_")]
    Ztr = clean_with_Z.loc[X_train_idx, cols].values
    Zte = clean_with_Z.loc[X_test_idx,  cols].values
    return Ztr, Zte

In [ ]:
# ──────────────────────────────────────────────────
# Simple‐avg embeddings (already broadcasted)
# ──────────────────────────────────────────────────
Ztr_s_pca,  Zte_s_pca  = get_train_test_Z(clean_with_Z_avg_pca,  "pca", X_train.index, X_test.index)
Ztr_s_lap,  Zte_s_lap  = get_train_test_Z(clean_with_Z_avg_lap,  "lap", X_train.index, X_test.index)
Ztr_s_diff, Zte_s_diff = get_train_test_Z(clean_with_Z_avg_diff, "diff", X_train.index, X_test.index)

# ──────────────────────────────────────────────────
# TF‑IDF embeddings (already broadcasted)
# ──────────────────────────────────────────────────
Ztr_t_pca,  Zte_t_pca  = get_train_test_Z(clean_with_Z_tf_pca,  "pca", X_train.index, X_test.index)
Ztr_t_lap,  Zte_t_lap  = get_train_test_Z(clean_with_Z_tf_lap,  "lap", X_train.index, X_test.index)
Ztr_t_diff, Zte_t_diff = get_train_test_Z(clean_with_Z_tf_diff, "diff", X_train.index, X_test.index)

# Now you have six (Ztr, Zte) pairs ready for regression:
methods = [
    ("Simple‑avg PCA",       Ztr_s_pca,  Zte_s_pca),
    ("Simple‑avg Laplacian", Ztr_s_lap,  Zte_s_lap),
    ("Simple‑avg Diffusion", Ztr_s_diff, Zte_s_diff),
    ("TF‑IDF PCA",           Ztr_t_pca,  Zte_t_pca),
    ("TF‑IDF Laplacian",     Ztr_t_lap,  Zte_t_lap),
    ("TF‑IDF Diffusion",     Ztr_t_diff, Zte_t_diff),
]


In [ ]:
print("Simple‑avg PCA shapes:", Ztr_s_pca.shape, Zte_s_pca.shape)

Simple‑avg PCA shapes: (8378, 3) (2793, 3)


In [ ]:
# SANITY CHECK

def check_nans_in_embeddings(methods):
    """
    Given a list of (name, Ztr, Zte) tuples, report NaN counts in each.
    Returns a dict mapping method names to {'Train': count, 'Test': count}.
    """
    results = {}
    for name, Ztr, Zte in methods:
        train_nans = np.isnan(Ztr).sum()
        test_nans  = np.isnan(Zte).sum()
        results[name] = {'Train': train_nans, 'Test': test_nans}
        print(f"\n{name}:")
        print(f"  Training NaNs: {train_nans}")
        print(f"  Testing  NaNs: {test_nans}")
    return results

# Call the function
nan_results = check_nans_in_embeddings(methods)


Simple‑avg PCA:
  Training NaNs: 0
  Testing  NaNs: 0

Simple‑avg Laplacian:
  Training NaNs: 0
  Testing  NaNs: 0

Simple‑avg Diffusion:
  Training NaNs: 0
  Testing  NaNs: 0

TF‑IDF PCA:
  Training NaNs: 0
  Testing  NaNs: 0

TF‑IDF Laplacian:
  Training NaNs: 0
  Testing  NaNs: 0

TF‑IDF Diffusion:
  Training NaNs: 0
  Testing  NaNs: 0


In [ ]:
# 3️⃣  Run OLS/Ridge regressions & report OOS R² and MSE
# -------------------------------------------------
results = []
for name, Ztr, Zte in methods:
    mdl  = LinearRegression().fit(Ztr, y_train)   # SWAP for Ridge(alpha=1) OR LinearRegression()
    pred = mdl.predict(Zte)
    results.append({
        "Method": name,
        "OOS R²": r2_score(y_test, pred),
        "MSE":    mean_squared_error(y_test, pred)
    })

results_df = pd.DataFrame(results).set_index("Method")
print(results_df)

                        OOS R²       MSE
Method                                  
Simple‑avg PCA       -0.000509  0.000563
Simple‑avg Laplacian -0.000260  0.000563
Simple‑avg Diffusion  0.000752  0.000563
TF‑IDF PCA           -0.001027  0.000564
TF‑IDF Laplacian     -0.000267  0.000563
TF‑IDF Diffusion     -0.000593  0.000563


In [ ]:
df = pd.DataFrame(results)

# 1⃣  Split Method into Variant & Technique
# We assume Method strings are "<Variant> <Technique>"
df[['Variant', 'Technique']] = df['Method'].str.split(' ', n=1, expand=True)

# 2⃣  Pivot so Technique is the index, Variant+Metric are columns
pivot = df.pivot(index='Technique', columns='Variant', values=['OOS R²', 'MSE'])

# 3⃣  Flatten and rename columns
# This will produce a MultiIndex like ('OOS R²','Simple‑avg'), etc.
# We want columns: ['S.Avg OOS R²','TFIDF OOS R²','S.Avg MSE','TFIDF MSE']
col_map = {
    'Simple‑avg': 'S.Avg',
    'TF‑IDF':     'TFIDF'
}

new_cols = []
for metric, variant in pivot.columns:
    short = col_map.get(variant, variant)
    new_cols.append(f"{short} {metric}")

pivot.columns = new_cols

# 4⃣  (Optional) order the columns as you like
pivot = pivot[['S.Avg OOS R²', 'TFIDF OOS R²', 'S.Avg MSE', 'TFIDF MSE']]

print(pivot)

# Save the pivoted DataFrame to a CSV file
pivot.to_csv('regression_results_TABLE.csv')

           S.Avg OOS R²  TFIDF OOS R²  S.Avg MSE  TFIDF MSE
Technique                                                  
Diffusion      0.000752     -0.000593   0.000563   0.000563
Laplacian     -0.000260     -0.000267   0.000563   0.000563
PCA           -0.000509     -0.001027   0.000563   0.000564


In [ ]:
# # (Optional) assemble a single design DataFrame
# # --------------------------------------------------
# # If you want to run a single regression that includes all three sets of factors:
# clean_all = (
#     clean_df
#     .join(clean_with_Z_pca[[f"pca_{i}" for i in range(1, 11)]])
#     .join(clean_with_Z_lap[[f"lap_{i}" for i in range(1, 11)]])
#     .join(clean_with_Z_diff[[f"diff_{i}" for i in range(1, 11)]])
# )

# # 4️⃣  (Optional) Combined regression
# # -----------------------------------
# Ztr_all = clean_all.loc[X_train.index,  pca_cols + lap_cols + diff_cols].values
# Zte_all = clean_all.loc[X_test.index,   pca_cols + lap_cols + diff_cols].values

# print("\n=== Combined factors regression ===")
# eval_model(Ztr_all, Zte_all, y_train, y_test, "All Factors")


# OLS REGRESSIONS ON 10 PCs

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>OOS R²</th>
      <th>MSE</th>
    </tr>
    <tr>
      <th>Method</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Laplacian</th>
      <td>-0.000267</td>
      <td>0.000563</td>
    </tr>
    <tr>
      <th>Diffusion</th>
      <td>-0.001862</td>
      <td>0.000564</td>
    </tr>
    <tr>
      <th>PCA</th>
      <td>-0.002961</td>
      <td>0.000565</td>
    </tr>
  </tbody>
</table>
</div>

# OLS REGRESSIONS ON 3 PCs

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>OOS R²</th>
      <th>MSE</th>
    </tr>
    <tr>
      <th>Method</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Laplacian</th>
      <td>-0.000267</td>
      <td>0.000563</td>
    </tr>
    <tr>
      <th>PCA</th>
      <td>-0.001027</td>
      <td>0.000564</td>
    </tr>
    <tr>
      <th>Diffusion</th>
      <td>-0.001124</td>
      <td>0.000564</td>
    </tr>
  </tbody>
</table>
</div>

# RIDGE REGRESSIONS ON 3 PCs
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>OOS R²</th>
      <th>MSE</th>
    </tr>
    <tr>
      <th>Method</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Laplacian</th>
      <td>-0.000268</td>
      <td>0.000563</td>
    </tr>
    <tr>
      <th>Diffusion</th>
      <td>-0.000631</td>
      <td>0.000563</td>
    </tr>
    <tr>
      <th>PCA</th>
      <td>-0.001026</td>
      <td>0.000564</td>
    </tr>
  </tbody>
</table>
</div>

# RIDGE REGRESSIONS ON 3 PCs w/ diffusion t = 5

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>OOS R²</th>
      <th>MSE</th>
    </tr>
    <tr>
      <th>Method</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Laplacian</th>
      <td>-0.000268</td>
      <td>0.000563</td>
    </tr>
    <tr>
      <th>Diffusion</th>
      <td>-0.000631</td>
      <td>0.000563</td>
    </tr>
    <tr>
      <th>PCA</th>
      <td>-0.001026</td>
      <td>0.000564</td>
    </tr>
  </tbody>
</table>
</div>